### PR TITLE
[TypeScript-Node] support setting the content-type header per-call

### LINKED ISSUE #1867: [REQ] [TypeScript-Node] support setting the content-type header per-call
### Is your feature request related to a problem? Please describe.

Some API endpoints accept multiple content-types with the same data format.  For example, many of the Kubernetes API PATCH endpoints accept JSON data format, but with several different possible content-types (*e.g.*, application/strategic-merge-patch+json, application/merge-patch+json, or application/json-patch+json).

The typescript-node implementation makes it cumbersome to call the same endpoint with different content-types, although you can accomplish it by setting content-type in `.defaultHeaders` before the call, and restoring the previous value of content-type after the call.

## Describe the solution you'd like

I'd like to specify this on a per-call basis.  For example, by adding a final `options` parameter to each call:

```js
const result = await client.patchNamespace('foo', body, { headers: { 'content-type': 'application/strategic-merge-patch+json' } })
```

## Describe alternatives you've considered

The workaround mentioned above:

```js
const contentType = client.defaultHeaders['content-type']
client.defaultHeaders['content-type'] = 'application/strategic-merge-patch+json'
const resultPromise = client.patchNamespace('foo', body)
client.defaultHeaders['content-type'] = contentType
const result = await resultPromise
```

## Additional context

https://github.com/swagger-api/swagger-codegen/issues/9035

### PR DESCRIPTION
**Pull request by silasbw**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This approach is inspired by the typescript-fetch implementation in swagger-codegen.

Fixes #1867

### CODE REVIEW

**Comment by wing328** [comment]
cc @TiFu (2017/07) @taxpon (2017/07) @sebastianhaas (2017/07) @kenisteward (2017/07) @Vrolijkx (2017/09) @macjohnny (2018/01) @nicokoenig (2018/09) @topce (2018/10)

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache [hunk 1]
```diff
@@ -127,7 +127,7 @@ export class {{classname}} {
      * @param {{paramName}} {{description}}
      {{/allParams}}
      */
-    public {{nickname}} ({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}{{#hasMore}}, {{/hasMore}}{{/allParams}}) : Promise<{ response: http.{{#supportsES6}}IncomingMessage{{/supportsES6}}{{^supportsES6}}ClientResponse{{/supportsES6}}; {{#returnType}}body: {{{returnType}}}; {{/returnType}}{{^returnType}}body?: any; {{/returnType}} }> {
+    public {{nickname}} ({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}options: any = {}) : Promise<{ response: http.{{#supportsES6}}IncomingMessage{{/supportsES6}}{{^supportsES6}}ClientResponse{{/supportsES6}}; {{#returnType}}body: {{{returnType}}}; {{/returnType}}{{^returnType}}body?: any; {{/returnType}} }> {
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]: I suggest to be more precise about the typing/naming what option we set:
```suggestion
    public {{nickname}} ({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}options: {headerParams: {[name: string]: value}} = {headerParams: {}}) : Promise<{ response: http.{{#supportsES6}}IncomingMessage{{/supportsES6}}{{^supportsES6}}ClientResponse{{/supportsES6}}; {{#returnType}}body: {{{returnType}}}; {{/returnType}}{{^returnType}}body?: any; {{/returnType}} }> {
```

File: modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache [hunk 2]
```diff
@@ -152,6 +152,7 @@ export class {{classname}} {
 {{#headerParams}}
         localVarHeaderParams['{{baseName}}'] = ObjectSerializer.serialize({{paramName}}, "{{{dataType}}}");
 {{/headerParams}}
+        (<any>Object).assign(localVarHeaderParams, options.headers);
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-node/api-single.mustache]:
```suggestion
        (<any>Object).assign(localVarHeaderParams, options.headerParams);
```

**Comment by silasbw** [comment]
@macjohnny applied your suggestions, ty.

About naming `.headerParams` vs `.headers`, one argument for user `.headers` is that a few popular libraries/APIs use the name `headers` for a similar purpose ([Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Supplying_your_own_request_object), [request](https://github.com/request/request#custom-http-headers), [axios](https://github.com/axios/axios#request-config)). Using `.headers` instead of `.headerParams` might be "the least surprising".

**Comment by wing328** [comment]
@silasbw thanks for the PR, which has been merged into master.

Is it correct to say that this is a breaking change (without fallback) as I saw the changes in the method signature?

**Comment by silasbw** [comment]
I don't think it's breaking, because the change includes a default value for argument it adds to the function signature..

**Comment by wing328** [comment]
@silasbw 👌 thanks for the explanation.

**Pull request merged**